event-store-plans.md

```diff
@@ -0,0 +1,174 @@
+# Event Store Improvements for V7
+
+This document is brainstorming for improvements to the Event Store functionality in Marten V7. This does not include features meant for
+"CritterStackPro" products, but does include some necessary baseline infrastructure for planned "CritterStackPro" features.
+
+In some cases, this document outlines "poor man's" subset strategies for features that we will build in a more robust (and inevitably more complicated) way 
+in "CritterStackPro"
+
+The goal is to start a discussion about what is and isn't going to be in scope for V7. When this document is "complete", the output is a new set
+of GitHub issues marked for the V7 release. 
+
+
+## Raise events from asynchronous projection updates
+
+I'm going to allow Oskar address this one:-)
+
+From my side, I think the `ISessionListener` approach we have today is weak and should be a little more "outbox'd" such that it only fires off when
+the transaction is complete. 
+
+For the Wolverine integration, I'd like to be able to enroll in Wolverine's outbox as part of processing a page of events in the daemon to send out messages using the 
+aggregation data as part of the outgoing message.
+
+More generically, we need a good way to emit new events that will be processed as appends within the same daemon transaction. This might be a little tricky and will
+require a spike.
+
+**Gotta do this in a way such that users can opt into whether the new events should be emitted during projection rebuilds**
+
+
+## Sharding the Event Store tables
+
+This is a [long standing issue](https://github.com/JasperFx/marten/issues/770) that we didn't get around to in the huge V4 release.
+
+Some notes:
+
+* My personal proposal is to use native PostgreSQL sharding on the `is_archived` column to effect a "hot storage, cold storage" approach. This should be paired with
+  advice to users to be more aggressive about archiving event streams that are obsolete/stale/not active as a way to improve throughput over all. The daemon and every
+  other place where we query events automatically uses `is_archived = false` filters anyway
```

Comment:
I feel this would encourage the use of `is_archived` far more than it is currently being used by marten consumers. Might need to consider implications for things like projections.

For example, I currently only use is_archived as a soft-delete that will eventually be hard deleted (as part of restreaming usually). If it becomes a "move your old stuff here" feature then I will have challenges with projection consistency. In particular I would want to make sure that a new projection I create is backfilled with archived values as well, or at least configure my projection so that can be possible.

Reply:
My thoughts as well. That's mostly inspired by Eventuous's "hot / cold storage" idea

Reply:
See my write-up on the log compaction process: 
- https://event-driven.io/en/gdpr_in_event_driven_architecture/#log-compaction
- https://www.youtube.com/watch?v=gG6DGmYKk4I
- https://www.eventstore.com/blog/keep-your-streams-short-temporal-modelling-for-fast-reads-and-optimal-data-retention

And EventStoreDB scavenging process docs: https://developers.eventstore.com/server/v23.10/operations.html#scavenging-events

TLDR, we should have a formal archival process that works as a background process. It has to be either on-demand or semi-automatic. We should ensure that all archived events are published and processed via async daemon.

The detaching partition should be the last step, first people should be able to move the events somewhere else depending on the strategy.

I think that eventually we should support all scenarios I described. The basic building blocks should be part of Marten. So archival, also partial archival of stream, tombstoning, and hosted service giving option to plug in custom archival logic. The more advanced and precise scenarios should be part of CritterStackPro.

Reply:
Big fan of this feature.
> We should ensure that all archived events are published and processed via async daemon.
Yes please. This is currently an issue with my applications being able to use archival. The async daemon doesn't pick up any event that was just archived, so I can't move the async projection into a new state (or delete that document) on that initial archival event.